### DISCUSSION TITLE
How can I get the image display size?

### DISCUSSION THREAD

**Discussion by luckymore** [mozilla/pdf.js#17919] [Q&A]
Attach (recommended) or Link to PDF file here:
https://arxiv.org/pdf/2404.01370.pdf

Configuration:
- Web browser and its version: chrome 123.0.6312.107 (x86_64)
- Operating system and its version: 13.6.3 (22G436)
- PDF.js version: '4.0.379'
- Is a browser extension: No

Steps to reproduce the problem:
1. Get the `viewport` with `page.getViewport({ scale: 1 })`
```
{
    "viewBox": [
        0,
        0,
        612,
        792
    ],
    "scale": 1,
    "rotation": 0,
    "offsetX": 0,
    "offsetY": 0,
    "transform": [
        1,
        0,
        0,
        -1,
        0,
        792
    ],
    "width": 612,
    "height": 792
}
```
2. Get the Image in page 2 with `page.objs.get(id)`
```
[
    {
        "width": 2000,
        "height": 2000,
        "kind": 2,
        "dataLen": 12000000,
        "ref": "203R"
    },
    {
        "width": 1749,
        "height": 1749,
        "kind": 2,
        "dataLen": 9177003,
        "ref": "204R"
    },
    {
        "width": 960,
        "height": 36,
        "kind": 2,
        "dataLen": 103680,
        "ref": "205R"
    }
]
```

What went wrong? 

Why all the images size greater than page size?
![image](https://github.com/mozilla/pdf.js/assets/5390013/7b2774f2-9e83-4f22-83a4-cd8b438ed5cd)

Link to a viewer (if hosted on a site other than mozilla.github.io/pdf.js or as Firefox/Chrome extension):

```[tasklist]
### Tasks
```

**Comment by carlbrn** [comment]
Here the method I use. Although, I'm not sure 'transform' is always happening before 'paintImageXObject'. So far it has been the case with the PDFs I use. 

Basically, if you iterate through the operator list you can catch the image 'transform' and use that.

```Typescript
type Transform = [number, number, number, number, number, number]

// Use the unit to convert numbers to inch if needed.
const unit = page.userUnit * 72

const opList = await page.getOperatorList()

let lastTransform: Transform | null = null

for (let i = 0; i < opList.fnArray.length; i++) {

	if (opList.fnArray[i] === pdfjsLib.OPS.transform) {
	    const args = opList.argsArray[i]
	    lastTransform = args as Transform
	}

	if (opList.fnArray[i] === pdfjsLib.OPS.paintImageXObject && lastTransform) {
		// here use the transform to get the size, you can also use it to get the position and polygon.
		const imageWidth = Math.abs(lastTransform[0])
		const imageHeight = Math.abs(lastTransform[3])

		const imageWidthInch = Math.abs(lastTransform[0]) / unit
		const imageHeightInch = Math.abs(lastTransform[3]) / unit
	}
}
```
